### ISSUE TITLE
SessionUpdated event can still fire after StopSession() is called

### ISSUE THREAD

**Issue by WikkidEdd** [Azure/azure-spatial-anchors-samples#356]
**Description**

Occasionally after calling SpatialAnchorManager.StopSession(), the SessionUpdated event will still get called. This has several implications.

Internally StopSession() sets sessionStatus to null. If OnSessionUpdate gets called afterwards it will set sessionStatus to the previous session status again. Consequently, both IsReadyForCreate and SessionStatus will report incorrect values for what is a stopped session.

Furthermore, if any client code responds to SessionUpdated assuming that it only ever fires when the session is running it can cause unexpected behaviour.

**Steps to reproduce the issue**

1. Create an ASA Session
2. Start ASA Session
3. Register for SessionUpdated and SesssionStopped events with functions to log out when they are fire
4. Create and upload anchor.
5. Call StopSession
6. Repeat steps 2 to 4.

You will eventually get a case where you see a SessionUpdate logged out after SessionStopped.

**Expected behavior**

After calling StopSession, the SessionUpdate event will not fire.

**Screenshots**

**Development information (please complete the following information)**

 - Development Platform\Version: Unity 2020.3.19f1
 - Azure Spatial Anchors SDK version: 2.12.0

**AR Device information (please complete the following information):**

 - Device: Hololens 2
 - OS: 10.0.20348.1506

**Comment by XiangyuL-Microsoft** [comment]
@WikkidEdd 

Hi, Edd. Is the issue happening with the same app as [issue 355](https://github.com/Azure/azure-spatial-anchors-samples/issues/355)? Is it possible for us to read your code for investigation purpose?

Thank you very much.

Best Regards

**Comment by WikkidEdd** [comment]
It was happening with the same app as #355, but this one is easily reproducible in it a test app. I can share the test app code, but you should be able to reproduce it using the steps I've provided.

I'll upload the test project shortly.

**Comment by WikkidEdd** [comment]
You can find the test project here https://github.com/fracturegames/ASABug_Repro

1. Clone it
2. Add your ASA API Keys etc.
3. Build Unity Project for UWP
4. Build VS project with Release and ARM64
5. Run the app

You will see a grey panel with some debug information. You will see it cycling through the repro steps and counting up the number of runs. Eventually you should see an error appear in read `SessionUpdated while session is not started`. You can then stop the app and check the logs to see that the SessionUpdated event has been fired since the StopSession function was called.

It can sometimes take 1-10 runs to trigger the problem, but can also be up to 50 as well. If you get up to 50, close the app and start again.

On one occasion while running this test, I also got the error 
`Unexpected error gathering feedback data`

**Comment by XiangyuL-Microsoft** [comment]
@WikkidEdd 

Hi, Edd

After investing the implementation of ASA Unity Plugin and ASA SDK, we would like to claim that the behavior which you reported as bug is by design.

The method [CloudSpatialAnchorSession.Stop](https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.spatialanchors.cloudspatialanchorsession.stop?view=spatialanchors-dotnet#microsoft-azure-spatialanchors-cloudspatialanchorsession-stop) stops capturing environment data for the session and cancels any outstanding locate operations. Environment data is maintained. There is no guarantee that the status of certain CloudSpatialAnchorSession would not be changed. Once any change happens, the [CloudSpatialAnchorSession.SessionUpdated](https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.spatialanchors.cloudspatialanchorsession.sessionupdated?view=spatialanchors-dotnet) would surely be triggered.

In other words, we shouldn't expect that CloudSpatialAnchorSession.SessionUpdated would not be triggered just because CloudSpatialAnchorSession.Stop was invoked.

**Comment by Varnauld** [comment]
Thank you @XiangyuL-Microsoft, 

Hello @WikkidEdd - was the explanation above helpful? Please let us know if this helps you resolve this issue, 

Thank you

**Comment by robbbbbb** [comment]
I'm @WikkidEdd 's colleague and have been helping him chase down issues we're seeing with the ASA SDK. 

Reading his original issue again I think you can see that it's not just a question of semantics of the order the Stop and the SessionUpdated happen in. 

I think it's reasonable for you to consider that correct, but I have to say, it seems wrong from the perspective of an SDK user. If _a session_ a stopped, then it doesn't make sense to us that it would still be updated. It just seems like a race condition.

Putting the semantics to one side, it's clear that the actual package code itself is responding to the SessionUpdated callback in an incorrect way _if the callback occurs when the session is stopped_. The specific error that @WikkidEdd calls out in his issue being:

> Internally StopSession() sets sessionStatus to null. If OnSessionUpdate gets called afterwards it will set sessionStatus to the previous session status again. Consequently, both IsReadyForCreate and SessionStatus will report incorrect values for what is a stopped session.

(note that SessionStatus is used for calculating the value of `IsReadyForCreate` - that property should be `false `if `sessionStatus `is null, instead if this race condition occurs it will likely end up `true `instead)

The code in question is in the ASA Unity package in SpatialAnchorManager.cs. Please take a look as we are very confident this is _not_ by design.

**Comment by XiangyuL-Microsoft** [comment]
@robbbbbb 

Hi, Rob

It seems that you have a little misunderstanding about the APIs.

1. If you don't need a session, just destroy it other than stop it. Once you destroyed a session, there wouldn't be any further update event triggered for it.

2. You are right, once a stopped session is updated, the "IsReadyForCreate" may become true again. I understand this may be what make your confusing. But the fact is that the start/stop doesn't determine if anchor is creatable or not. The start/stop only control capturing of environment data.

3. About the race condition you mentioned, I think we actually need to consider if the reset of sessionStatus during StopSession is correct. I would discuss with my colleague and follow up with this issue.

Thank you very much.

Best Regards

**Comment by XiangyuL-Microsoft** [comment]
@WikkidEdd @robbbbbb 

Hi, Edd & Rob

After discussing, we decided to modify the Unity Plugin to avoid certain confuse. The patch would be released with the next version of SDK.

**Comment by XiangyuL-Microsoft** [comment]
@WikkidEdd 

We just released a new version of SDK [2.13.0](https://github.com/Azure/azure-spatial-anchors-samples/releases/tag/v2.13.0).
Could you try this new version of SDK to see if the problem still persists please?
Thank you very much.

**Comment by Varnauld** [comment]
Hello @WikkidEdd - did the last update in July resolve the issues you were seeing? If yes, we will go ahead and close this but if you have any further questions or concerns, please let us know. Thank you
